File: stage3-translator/src/symbols/Environment.java
package symbols;

import errors.SemanticError;
import java.util.Hashtable;

public class Environment {
	private Hashtable<String, Double> table;
	private Environment previous;
	
	public Environment() {
		table = new Hashtable<String, Double>();
		previous = null;
	}
	
	public Environment(Environment prev) {
		table = new Hashtable<String, Double>();
		previous = prev;
	}
	
	public Environment getPrevious() {
		return previous;
	}
	
	public void setPrevious(Environment previous) {
		this.previous = previous;
	}
	
	public void insert(String varName) throws SemanticError{
		if (table.get(varName) == null) {
			table.put(varName, Double.valueOf(0.0));
		} else {
			throw new SemanticError("The variable was already defined");
		}
	}
	
	public Double lookup(String varName) {
		Double found;
		for (Environment env = this; env.previous != null; env = env.previous) {
			found = env.table.get(varName);
			if (found != null) {
				return found;
			}
		}
		return null;
	}
	
	public void set(String varName, double value) throws SemanticError{
		Double found = null;
		for (Environment env = this; env.previous != null; env = env.previous) {
			found = env.table.get(varName);
			if (found != null) {
				env.table.put(varName, Double.valueOf(value));
				break;
			}
		}
		if (found == null) {
			throw new SemanticError("The variable has not been defined");
		}
	}
}

File: stage-3-interpreter/Environment.py
class Environment:
    __table = {}
    __previous = None

    def __init__(self, env = None):
        self.__previous = env
        self.__table = {}

    def getPrevious(self):
        return self.__previous
    
    def lookup(self, variable):
        if variable in self.__table:
            # The variable is a local to this scope.
            return self.__table[variable]
        elif self.__previous != None:
            # The variable maybe is declared in an upper scope.
            return self.__previous.lookup(variable)
        else: 
            return None
    
    def insert(self, variable):
        if not variable in self.__table:
            # the variable has not been declared.
            self.__table[variable] = (None, None)
            return True
        else:
            # the variable has already been declared.
            return False
        
    def set(self, variable, type = None, value = None):
        env = self
        found = False
        while env != None:
            if variable in env.__table:
                found = True
                break
            env = env.__previous
        
        if found:
            # the variable is declared.
            env.__table[variable] = (type, value)
            return True
        else: 
            # the variable is not declared.
            return False
